File: stackunderflow/adapters/kiro.py
"""Kiro (kiroagent) session adapter.

Reads chat files Kiro writes under VS Code-style globalStorage:

- macOS: ``~/Library/Application Support/Kiro/User/globalStorage/kiro.kiroagent/``
- Windows: ``%APPDATA%\\Kiro\\User\\globalStorage\\kiro.kiroagent\\``  *(untested)*
- Linux: ``~/.config/Kiro/User/globalStorage/kiro.kiroagent/``  *(untested)*

Each ``.chat`` file is a single JSON document::

    {
      "executionId": "...",
      "actionId": "...",
      "chat": [
        {"role": "human" | "bot" | "tool", "content": "..."}
      ],
      "metadata": {
        "modelId": "claude.3.5.sonnet",
        "startTime": "...",
        "endTime": "...",
        "workflowId": "..."
      }
    }

**Quirks**:

- **Tokens are estimated** from content length / 4 (Kiro doesn't record
  per-call usage). We mark every emitted Record with
  ``raw["cost_source"] = "estimated"`` so the cost layer can decide to
  show or down-weight these numbers.
- Model ids come dot-separated (``claude.3.5.sonnet``); we normalise
  them to the dash-separated form (``claude-3-5-sonnet``) so the
  Anthropic pricer's family heuristic matches.
- One Record per execution (the entire chat is rolled up into a single
  assistant turn). Resumable reads (``since_offset``) are by event index
  — Kiro files are small and aren't streamed.

Spec §3 (multi-provider).

Defensive sizing: ``.chat`` files larger than ``MAX_SESSION_FILE_BYTES``
(128 MB; see ``stackunderflow/adapters/_streaming.py``) are **skipped
with a logged warning** rather than parsed. Single-document JSON cannot
stream, so the cap is the only safety net."""

from __future__ import annotations

import json
import logging
import os
import sys
from collections.abc import Iterator
from pathlib import Path

from ._streaming import stat_or_skip
from .base import Record, SessionRef

_log = logging.getLogger(__name__)


def _kiro_global_storage() -> Path:
    """Return the platform-appropriate Kiro ``globalStorage`` root.

    Same platform-branch shape as the VS Code adapters; ``APPDATA`` is
    read at call time so tests can monkeypatch it. Real-box validation on
    Windows/Linux is still pending (catalog §3).
    """
    if sys.platform.startswith("win"):
        return Path(os.environ.get("APPDATA", "")) / "Kiro" / "User" / "globalStorage" / "kiro.kiroagent"
    if sys.platform.startswith("linux"):
        return Path.home() / ".config" / "Kiro" / "User" / "globalStorage" / "kiro.kiroagent"
    return Path.home() / "Library" / "Application Support" / "Kiro" / "User" / "globalStorage" / "kiro.kiroagent"


_DEFAULT_MODEL = "kiro-auto"


class KiroAdapter:
    """Source adapter for the Kiro agent extension."""

    name = "kiro"

    def __init__(self, storage_root: Path | None = None) -> None:
        self._root = storage_root or _kiro_global_storage()

    # ── enumeration ───────────────────────────────────────────────────

    def source_roots(self) -> list[Path]:
        """Roots ``backup create`` copies — the same data
        ``enumerate()`` reads. Self-declared here (like ``name``),
        never listed centrally.
        """
        return [self._root]

    def enumerate(self) -> Iterator[SessionRef]:
        root = self._root
        if not root.is_dir():
            return

        # Kiro stores ``.chat`` files at the storage root and possibly
        # under nested directories (workspace-scoped subtrees). Walk the
        # tree and yield every ``*.chat`` we find.
        for fp in sorted(root.rglob("*.chat")):
            try:
                stat = fp.stat()
            except OSError as exc:
                _log.warning("Cannot stat Kiro chat %s: %s", fp, exc)
                continue

            session_id = fp.stem  # default; refined from metadata in read()
            workflow_id, project_slug = _peek_metadata(fp)
            if workflow_id:
                session_id = workflow_id

            yield SessionRef(
                provider=self.name,
                project_slug=project_slug or "kiro",
                session_id=session_id,
                file_path=fp,
                file_mtime=stat.st_mtime,
                file_size=stat.st_size,
                source_kind="file",
                source_hint=None,
            )

    # ── reading ───────────────────────────────────────────────────────

    def read(self, ref: SessionRef, *, since_offset: int = 0) -> Iterator[Record]:
        # Single-document JSON: must call ``json.load`` on the whole
        # file, so the only safety net is the defensive 128 MB cap.
        # Above the cap we yield nothing — same convention as the JSONL
        # adapters.
        if stat_or_skip(ref.file_path) is None:
            return
        try:
            with ref.file_path.open("rb") as fh:
                data = json.load(fh)
        except (OSError, json.JSONDecodeError, ValueError) as exc:
            _log.warning("Cannot read Kiro chat %s: %s", ref.file_path, exc)
            return
        if not isinstance(data, dict):
            return

        meta = data.get("metadata") or {}
        if not isinstance(meta, dict):
            meta = {}

        raw_model = meta.get("modelId")
        model = _normalize_model(raw_model) if isinstance(raw_model, str) else _DEFAULT_MODEL

        timestamp = str(meta.get("startTime") or meta.get("endTime") or "")

        chat = data.get("chat") or []
        if not isinstance(chat, list):
            chat = []

        # Kiro rolls up an entire execution into one logical assistant
        # turn. We compute a single seq for that turn (event index 0)
        # and emit one Record. ``since_offset >= 0`` skips it so the
        # resume contract still holds (a "midpoint" past the only record
        # yields nothing).
        if since_offset > 0:
            return

        human_text, bot_text = _join_chat(chat)
        # Token estimate: content_chars // 4.
        input_tokens = max(len(human_text) // 4, 0)
        output_tokens = max(len(bot_text) // 4, 0)

        # Mark estimated so the cost layer can flag / discount.
        raw_payload = dict(data)
        raw_payload["cost_source"] = "estimated"

        yield Record(
            provider=self.name,
            session_id=ref.session_id,
            seq=0,
            timestamp=timestamp,
            role="assistant",
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cache_create_tokens=0,
            cache_read_tokens=0,
            content_text=bot_text,
            tools=_extract_tools(bot_text),
            cwd=None,
            is_sidechain=False,
            uuid=str(data.get("executionId") or f"{ref.session_id}:0"),
            parent_uuid=None,
            raw=raw_payload,
        )


# ── helpers ───────────────────────────────────────────────────────────


def _peek_metadata(fp: Path) -> tuple[str, str]:
    """Return ``(workflow_id, project_slug)`` from the file's metadata block.

    Both default to empty when the file is unreadable or malformed.
    Project slug currently comes from the parent directory name as a
    proxy for "workspace" — Kiro's workspace-hash → directory mapping is
    not exposed in v1.
    """
    try:
        with fp.open("rb") as fh:
            obj = json.load(fh)
    except (OSError, json.JSONDecodeError, ValueError):
        return "", ""
    if not isinstance(obj, dict):
        return "", ""
    meta = obj.get("metadata") or {}
    workflow_id = meta.get("workflowId") if isinstance(meta, dict) else None
    return (
        str(workflow_id or ""),
        fp.parent.name or "kiro",
    )


def _normalize_model(model_id: str) -> str:
    """``claude.3.5.sonnet`` → ``claude-3-5-sonnet``.

    Dots become dashes; everything else passes through. Empty input
    falls back to ``kiro-auto``.
    """
    if not model_id:
        return _DEFAULT_MODEL
    return model_id.replace(".", "-").strip() or _DEFAULT_MODEL


def _join_chat(chat: list) -> tuple[str, str]:
    """Concatenate human-side and bot-side messages separately."""
    human_pieces: list[str] = []
    bot_pieces: list[str] = []
    for entry in chat:
        if not isinstance(entry, dict):
            continue
        role = entry.get("role")
        content = entry.get("content")
        text = content if isinstance(content, str) else ""
        if role == "human":
            human_pieces.append(text)
        elif role == "bot":
            bot_pieces.append(text)
        # ``tool`` role is ignored for token estimation; tool *names*
        # are extracted from bot text via _extract_tools.
    return "\n".join(human_pieces), "\n".join(bot_pieces)


def _extract_tools(bot_text: str) -> tuple[str, ...]:
    """Pull tool names from ``<tool_use><name>X</name>`` markers.

    Kiro embeds tool invocations as XML-style fragments inside bot text.
    We do a permissive split rather than a full XML parse — invalid/partial
    fragments yield zero tool names rather than raising.
    """
    if not bot_text:
        return ()
    out: list[str] = []
    cursor = 0
    while True:
        start = bot_text.find("<tool_use>", cursor)
        if start < 0:
            break
        nopen = bot_text.find("<name>", start)
        if nopen < 0:
            break
        nclose = bot_text.find("</name>", nopen)
        if nclose < 0:
            break
        name = bot_text[nopen + len("<name>") : nclose].strip()
        if name:
            out.append(name)
        cursor = nclose + len("</name>")
    return tuple(out)

File: rust/crates/stax-adapters/src/kiro.rs
//! Kiro (kiroagent) — the port of `stackunderflow/adapters/kiro.py`.
//!
//! Chat files live under a VS Code-style `globalStorage` root:
//!
//! ```text
//! macOS    ~/Library/Application Support/Kiro/User/globalStorage/kiro.kiroagent/
//! Linux    ~/.config/Kiro/User/globalStorage/kiro.kiroagent/          (untested)
//! Windows  %APPDATA%\Kiro\User\globalStorage\kiro.kiroagent\          (untested)
//! ```
//!
//! Each `.chat` file is one JSON document holding a whole execution, and the
//! adapter rolls that execution up into **one** assistant record.
//!
//! Three quirks that are not obvious from the shape:
//!
//! 1. **Tokens are estimated**, `len(text) // 4` over the human and bot sides
//!    separately, because Kiro records no usage. Every record carries
//!    `raw["cost_source"] = "estimated"` so the cost layer can down-weight it.
//!    `len` there is Python's — **code points, not bytes** — which is why this
//!    port counts `chars()`.
//! 2. **Model ids are dot-separated** (`claude.3.5.sonnet`) and are rewritten to
//!    the dash form so the Anthropic pricer's family heuristic matches.
//! 3. **Tool names are scraped out of the bot text** as `<tool_use><name>X</name>`
//!    fragments — a permissive split, never an XML parse, so a truncated
//!    fragment yields no tools instead of an error.
//!
//! Resume is by event index over a single record: `since_offset > 0` yields
//! nothing, which is what "past the only record" means.

use std::ffi::{OsStr, OsString};
use std::path::{Path, PathBuf};

use serde_json::{Map, Value};

use crate::base::{Record, SessionRef, SourceAdapter, Speed, stat_ref_fields};
use crate::jsonl;
use crate::pyval;
use crate::walk;

/// The provider key.
pub const NAME: &str = "kiro";

/// The model stamped when `metadata.modelId` is missing or blank
/// (`_DEFAULT_MODEL`).
pub const DEFAULT_MODEL: &str = "kiro-auto";

/// The extension-scoped directory every layout ends in.
pub const STORAGE_LEAF: &str = "kiro.kiroagent";

/// Kiro's `globalStorage` root, with the platform and environment injected
/// (`_kiro_global_storage`).
///
/// `os` is `std::env::consts::OS`; the Python original branches on
/// `sys.platform`, and the two agree on the only distinction that matters here
/// (windows / linux / everything-else-is-macOS). Injected rather than read so
/// the Windows and Linux layouts are testable on any box — the same
/// `sys.platform` test pattern the Python suite uses.
#[must_use]
pub fn resolve_global_storage(os: &str, appdata: Option<&OsStr>, home: Option<&Path>) -> PathBuf {
    let tail = |base: PathBuf| {
        base.join("Kiro")
            .join("User")
            .join("globalStorage")
            .join(STORAGE_LEAF)
    };
    if os == "windows" {
        // `Path(os.environ.get("APPDATA", ""))` — an unset APPDATA yields a
        // relative path, and `.is_dir()` then quietly fails. Ported as-is.
        return tail(PathBuf::from(appdata.unwrap_or_else(|| OsStr::new(""))));
    }
    let home = home.map_or_else(PathBuf::new, Path::to_path_buf);
    if os == "linux" {
        return tail(home.join(".config"));
    }
    tail(home.join("Library").join("Application Support"))
}

/// The Kiro source adapter (`KiroAdapter`).
#[derive(Debug, Clone)]
pub struct KiroAdapter {
    root: PathBuf,
}

impl Default for KiroAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl KiroAdapter {
    /// The platform-appropriate storage root, from the live environment.
    #[must_use]
    pub fn new() -> Self {
        #[allow(
            deprecated,
            reason = "std::env::home_dir is the platform-correct answer on the \
            1.97.1 pin; the claude and codex adapters carry the same allow"
        )]
        let home = std::env::home_dir();
        Self::with_env(std::env::consts::OS, std::env::var_os("APPDATA"), home)
    }

    /// Inject the platform and both environment inputs.
    #[must_use]
    pub fn with_env(os: &str, appdata: Option<OsString>, home: Option<PathBuf>) -> Self {
        Self {
            root: resolve_global_storage(os, appdata.as_deref(), home.as_deref()),
        }
    }

    /// Inject the storage root directly — `KiroAdapter(storage_root=…)`.
    #[must_use]
    pub fn with_storage_root(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }

    /// The storage root this adapter reads.
    #[must_use]
    pub fn storage_root(&self) -> &Path {
        &self.root
    }
}

impl SourceAdapter for KiroAdapter {
    fn name(&self) -> &str {
        NAME
    }

    fn enumerate(&self) -> Vec<SessionRef> {
        if !self.root.is_dir() {
            return Vec::new();
        }
        let mut out = Vec::new();
        // `.chat` files sit at the storage root and under workspace-scoped
        // subtrees, so the walk is recursive.
        for path in walk::rglob_suffix(&self.root, ".chat") {
            // Python warns and continues on a stat failure.
            let Some((mtime, size)) = stat_ref_fields(&path) else {
                continue;
            };
            let (workflow_id, project_slug) = peek_metadata(&path);
            let session_id = if workflow_id.is_empty() {
                walk::file_stem(&path)
            } else {
                workflow_id
            };
            let project_slug = if project_slug.is_empty() {
                NAME.to_string()
            } else {
                project_slug
            };
            out.push(SessionRef::file(
                NAME,
                project_slug,
                session_id,
                path,
                mtime,
                size,
            ));
        }
        out
    }

    fn read_into(&self, session: &SessionRef, since_offset: i64, sink: &mut dyn FnMut(Record)) {
        // Single-document JSON cannot stream, so the 128 MB cap is the only
        // safety net; above it we yield nothing, like every JSONL adapter.
        if jsonl::stat_or_skip(&session.file_path).is_none() {
            return;
        }
        // LOG: python warns "Cannot read Kiro chat %s".
        let Ok(text) = std::fs::read(&session.file_path) else {
            return;
        };
        let Some(document) = jsonl::parse_json(&text) else {
            return;
        };
        let Some(data) = document.as_object() else {
            return;
        };

        let empty = Map::new();
        // `data.get("metadata") or {}` then an isinstance check: falsy *and*
        // non-dict both become the empty mapping.
        let meta = data
            .get("metadata")
            .filter(|value| pyval::py_truthy(value))
            .and_then(Value::as_object)
            .unwrap_or(&empty);

        let model = match meta.get("modelId") {
            Some(Value::String(raw)) => normalize_model(raw),
            // A non-string modelId never reaches `_normalize_model`.
            _ => DEFAULT_MODEL.to_string(),
        };
        let timestamp = meta
            .get("startTime")
            .filter(|value| pyval::py_truthy(value))
            .or_else(|| meta.get("endTime").filter(|value| pyval::py_truthy(value)))
            .map_or_else(String::new, pyval::py_str);
        let chat = data
            .get("chat")
            .filter(|value| pyval::py_truthy(value))
            .and_then(Value::as_array)
            .map_or(&[][..], Vec::as_slice);

        // One logical turn per execution: a watermark past it yields nothing.
        if since_offset > 0 {
            return;
        }

        let (human_text, bot_text) = join_chat(chat);
        let mut raw_payload = data.clone();
        raw_payload.insert("cost_source".to_string(), Value::from("estimated"));

        sink(Record {
            provider: NAME.to_string(),
            session_id: session.session_id.clone(),
            seq: 0,
            timestamp,
            role: "assistant".to_string(),
            model: Some(model),
            input_tokens: estimate_tokens(&human_text),
            output_tokens: estimate_tokens(&bot_text),
            cache_create_tokens: 0,
            cache_read_tokens: 0,
            tools: extract_tools(&bot_text),
            content_text: bot_text,
            cwd: None,
            is_sidechain: false,
            uuid: data
                .get("executionId")
                .filter(|value| pyval::py_truthy(value))
                .map_or_else(|| format!("{}:0", session.session_id), pyval::py_str),
            parent_uuid: None,
            raw: Value::Object(raw_payload),
            speed: Speed::Standard,
        });
    }

    /// The storage root (`source_roots`). Kiro declares no `watch_paths`.
    fn source_roots(&self) -> Vec<PathBuf> {
        vec![self.root.clone()]
    }
}

/// `(workflow_id, project_slug)` from the document's metadata (`_peek_metadata`).
///
/// The project slug is the parent directory name, a stand-in for "workspace":
/// Kiro's workspace-hash → directory mapping is not exposed.
fn peek_metadata(path: &Path) -> (String, String) {
    let Ok(text) = std::fs::read(path) else {
        return (String::new(), String::new());
    };
    let Some(document) = jsonl::parse_json(&text) else {
        return (String::new(), String::new());
    };
    let Some(map) = document.as_object() else {
        return (String::new(), String::new());
    };
    let workflow_id = map
        .get("metadata")
        .filter(|value| pyval::py_truthy(value))
        .and_then(Value::as_object)
        .and_then(|meta| meta.get("workflowId"))
        .filter(|value| pyval::py_truthy(value))
        .map_or_else(String::new, pyval::py_str);
    let parent = path.parent().map_or_else(String::new, walk::dir_name);
    let parent = if parent.is_empty() {
        NAME.to_string()
    } else {
        parent
    };
    (workflow_id, parent)
}

/// `claude.3.5.sonnet` → `claude-3-5-sonnet` (`_normalize_model`).
#[must_use]
pub fn normalize_model(model_id: &str) -> String {
    if model_id.is_empty() {
        return DEFAULT_MODEL.to_string();
    }
    let normalized = model_id.replace('.', "-");
    let trimmed = normalized.trim();
    if trimmed.is_empty() {
        DEFAULT_MODEL.to_string()
    } else {
        trimmed.to_string()
    }
}

/// `(human_text, bot_text)`, each side joined with newlines (`_join_chat`).
///
/// The `tool` role is ignored for the token estimate; tool *names* come out of
/// the bot text instead.
fn join_chat(chat: &[Value]) -> (String, String) {
    let mut human: Vec<&str> = Vec::new();
    let mut bot: Vec<&str> = Vec::new();
    for entry in chat {
        let Some(map) = entry.as_object() else {
            continue;
        };
        // `content if isinstance(content, str) else ""` — a non-string content
        // still contributes an empty piece, and so still costs a newline.
        let text = map.get("content").and_then(Value::as_str).unwrap_or("");
        match map.get("role").and_then(Value::as_str) {
            Some("human") => human.push(text),
            Some("bot") => bot.push(text),
            _ => {}
        }
    }
    (human.join("\n"), bot.join("\n"))
}

/// `max(len(text) // 4, 0)` — Python's `len`, i.e. code points.
fn estimate_tokens(text: &str) -> i64 {
    i64::try_from(text.chars().count() / 4).unwrap_or(i64::MAX)
}

/// Tool names from `<tool_use><name>X</name>` markers (`_extract_tools`).
///
/// Deliberately a permissive scan: the `<name>` is searched for *after* the
/// `<tool_use>` marker with no bound, so a malformed fragment simply finds the
/// next well-formed one, and a missing terminator ends the scan rather than
/// raising.
#[must_use]
pub fn extract_tools(bot_text: &str) -> Vec<String> {
    if bot_text.is_empty() {
        return Vec::new();
    }
    let mut out = Vec::new();
    let mut cursor = 0_usize;
    while let Some(start) = bot_text[cursor..].find("<tool_use>").map(|at| cursor + at) {
        let Some(open) = bot_text[start..].find("<name>").map(|at| start + at) else {
            break;
        };
        let Some(close) = bot_text[open..].find("</name>").map(|at| open + at) else {
            break;
        };
        let name = bot_text[open + "<name>".len()..close].trim();
        if !name.is_empty() {
            out.push(name.to_string());
        }
        cursor = close + "</name>".len();
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn every_platform_layout_resolves_without_running_on_it() {
        let home = Path::new("/home/me");
        assert_eq!(
            resolve_global_storage("linux", None, Some(home)),
            Path::new("/home/me/.config/Kiro/User/globalStorage/kiro.kiroagent")
        );
        assert_eq!(
            resolve_global_storage("macos", None, Some(home)),
            Path::new(
                "/home/me/Library/Application Support/Kiro/User/globalStorage/kiro.kiroagent"
            )
        );
        assert_eq!(
            resolve_global_storage("windows", Some(OsStr::new(r"C:\Users\me\AppData")), None),
            Path::new(r"C:\Users\me\AppData")
                .join("Kiro")
                .join("User")
                .join("globalStorage")
                .join("kiro.kiroagent")
        );
        // An unset APPDATA yields a relative path, exactly as Python's
        // `Path(os.environ.get("APPDATA", ""))` does.
        assert_eq!(
            resolve_global_storage("windows", None, Some(home)),
            Path::new("Kiro/User/globalStorage/kiro.kiroagent")
        );
    }

    #[test]
    fn model_ids_lose_their_dots_and_blank_ones_fall_back() {
        assert_eq!(normalize_model("claude.3.5.sonnet"), "claude-3-5-sonnet");
        assert_eq!(normalize_model("gpt-5"), "gpt-5");
        assert_eq!(normalize_model(""), DEFAULT_MODEL);
        assert_eq!(normalize_model("   "), DEFAULT_MODEL);
        assert_eq!(normalize_model("  a.b  "), "a-b");
    }

    #[test]
    fn tool_scraping_survives_truncated_fragments() {
        assert_eq!(
            extract_tools("x<tool_use><name>Edit</name>y<tool_use><name>Read</name>"),
            vec!["Edit", "Read"]
        );
        // No marker, no name, no terminator: three ways to find nothing.
        assert!(extract_tools("plain text").is_empty());
        assert!(extract_tools("<tool_use>no name here").is_empty());
        assert!(extract_tools("<tool_use><name>unterminated").is_empty());
        assert!(extract_tools("<tool_use><name>  </name>").is_empty());
        assert!(extract_tools("").is_empty());
        // The name search is unbounded on purpose — a stray marker picks up
        // the next well-formed fragment rather than aborting the scan.
        assert_eq!(
            extract_tools("<tool_use>junk<name>Bash</name>"),
            vec!["Bash"]
        );
    }

    #[test]
    fn token_estimates_count_code_points_not_bytes() {
        // Eight ASCII characters -> 2; eight emoji (4 bytes each) -> also 2.
        assert_eq!(estimate_tokens("abcdefgh"), 2);
        assert_eq!(estimate_tokens("🙂🙂🙂🙂🙂🙂🙂🙂"), 2);
        assert_eq!(estimate_tokens(""), 0);
        assert_eq!(estimate_tokens("abc"), 0);
    }

    #[test]
    fn chat_sides_are_joined_separately_and_tool_turns_ignored() {
        let chat = json!([
            {"role": "human", "content": "ask"},
            {"role": "bot", "content": "answer"},
            {"role": "tool", "content": "ignored"},
            {"role": "bot", "content": 7},
            "not a dict",
        ]);
        let (human, bot) = join_chat(chat.as_array().expect("array"));
        assert_eq!(human, "ask");
        // The non-string bot content contributes an empty piece, and the
        // newline it costs is the observable difference.
        assert_eq!(bot, "answer\n");
    }

    #[test]
    fn an_absent_root_enumerates_empty_rather_than_failing() {
        let adapter = KiroAdapter::with_storage_root("/nonexistent/stax/kiro");
        assert!(adapter.enumerate().is_empty());
        assert_eq!(adapter.source_roots().len(), 1);
        assert!(adapter.watch_paths().is_empty());
    }
}
